geo/src/algorithm/voronoi_diagram.rs

```diff
@@ -0,0 +1,970 @@
+    f64: From<T>,
+{
+    let mut neighbours: Vec<(Option<usize>, Option<usize>)> = Vec::new();
+    let mut shared_edges: Vec<Line<T>> = Vec::new();
+
+    // Search the delaunay triangles for neighbour triangles and shared edges
+    for (tri_idx, tri) in triangles.iter().enumerate() {
+        for (other_idx, other_tri) in triangles.iter().enumerate() {
+            if tri_idx == other_idx {
+                continue;
+            }
+
+            if let Some(shared_edge) = tri.shares_edge(other_tri) {
+                if !neighbours.contains(&(Some(tri_idx), Some(other_idx)))
+                    && !neighbours.contains(&(Some(other_idx), Some(tri_idx)))
+                {
+                    neighbours.push((Some(tri_idx), Some(other_idx)));
+                }
+
+                let flipped_edge = Line::new(shared_edge.end, shared_edge.start);
+
+                if !shared_edges.contains(&shared_edge) && !shared_edges.contains(&flipped_edge) {
+                    shared_edges.push(shared_edge);
+                }
+            }
+        }
+    }
+
+    // For Voronoi diagrams, the triangles / circumcenters that are on the edge of the
+    // diagram require connections to infinity to ensure separation of points between
+    // voronoi cells.
+    // Voronoi cells on the borders can have 2 connections to infinity
+    // These connections to infinity will be bounded later
+    // Add the connections from infinity
+    let mut num_neighbours: Vec<usize> = triangles.iter().map(|_| 0).collect();
+    neighbours.iter().for_each(|x| {
+        // unwrap here is safe as neighbours do not contain None values yet
+        for n in [x.0.unwrap(), x.1.unwrap()] {
+            num_neighbours[n] += 1;
+        }
+    });
+
+    num_neighbours.iter().enumerate().for_each(|(idx, val)| {
+        for _ in 0..3 - val {
+            neighbours.push((None, Some(idx)));
+        }
+    });
+
+    Neighbours {
+        neighbours,
+        shared_edges,
+    }
+}
+
+fn get_perpendicular_line<T: GeoFloat>(line: &Line<T>) -> Result<Line<T>>
+where
+    f64: From<T>,
+{
+    let slope = f64::from(line.slope());
+
+    // Vertical line
+    if slope.is_infinite() {
+        Ok(Line::new(
+            coord! {x: line.start.x, y: line.start.y},
+            coord! {x: line.start.x - line.dy(), y: line.start.y},
+        ))
+    } else if slope == 0. {
+        Ok(Line::new(
+            coord! {x: line.start.x, y: line.start.y},
+            coord! {x: line.start.x, y: line.start.y + line.dx()},
+        ))
+    } else {
+        let midpoint = coord! {
+            x: f64::from(line.start.x) + f64::from(line.dx()) / 2.,
+            y: f64::from(line.start.y) + f64::from(line.dy()) / 2.,
+        };
+        let m = -1. / slope;
+        // y = mx + b
+        let b = midpoint.y - m * midpoint.x;
+        let x = midpoint.x + f64::from(line.dx());
+        let y = m * x + b;
+        let x = T::from(x).ok_or(VoronoiDiagramError::CannotConvertBetweenGeoGenerics)?;
```

Comment:
add a blank line above this, as the comment a few lines above is about `let y = ..`

Reply:
also this `x` and `y` looks like it is `end_x` & `end_y`, so would be nice if it is named as such, and possibly appears after `start_x` and `start_y`